Add exercise: state-of-tic-tac-toe

SaschaMann opened the pull request:
This is based on the research exercise julia-1-a

Extensions of this might be possible in the future, hence tic-tac-toe may not be ideal as a slug, so I'm going with state-of-tic-tac-toe.

SaschaMann commented:
/format

SaschaMann commented:
@ErikSchierboom the order of this is all wrong but I think you have a script to fix that, guessing based on #1960? (perhaps that should be part of `/format`, too)

SaschaMann commented:
/format

SaschaMann commented:
These arrays should be multi-line as in #1968.

@petertseng could you perhaps run your script on this PR? (feel free to push to the branch)

SaschaMann commented:
Some example solutions in Julia:

```julia
function gamestate(bs)
    # Parse to Matrix
    b = hcat(split.(split(bs, "\n"), "")...)

    # Invalid boards
    nₒ = count(x -> x == "O", b)
    nₓ = count(x -> x == "X", b)
    # Not taking turns in order 
    if nₒ > nₓ || abs(nₒ - nₓ) > 1
        throw(DomainError(b, "Invalid board: did not take turns in order"))
    end

    wins = 0

    for i in 1:3
        # Row victory
        wins += b[i, 1] != " " && (b[i, 1] == b[i, 2] == b[i, 3]) ? 1 : 0
        
        # Col victory
        wins += b[1, i] != " " && (b[1, i] == b[2, i] == b[3, i]) ? 1 : 0
    end

    # Diagonal victories
    wins += b[1, 1] != " " && (b[1, 1] == b[2, 2] == b[3, 3]) ? 1 : 0
    wins += b[1, 3] != " " && (b[1, 3] == b[2, 2] == b[3, 1]) ? 1 : 0

    if wins == 1
        return :win
    elseif wins > 1
        throw(DomainError(b, "Invalid board: more than one win"))
    end

    # Draw
    " " ∉ b && return :draw

    :ongoing
end
```

```julia
using LinearAlgebra

function gamestate(bs)
    # Parse to Int Matrix
    b_str = hcat(split.(split(bs, "\n"), "")...)
    replace!(b_str, "X" => "1")
    replace!(b_str, " " => "0")
    replace!(b_str, "O" => "-1")
    b = parse.(Int, b_str)

    # It's a win if any of the rows, columns, or diagonals add up to ±3
    wins = count(sum(row) in (-3, 3) for row in eachrow(b)) + count(sum(col) in (-3, 3) for col in eachcol(b)) + (sum(diag(b)) in (-3, 3) ? 1 : 0) + (sum(diag(rotr90(b))) in (-3, 3) ? 1 : 0)
    if wins == 1
        :win
    elseif wins > 1
        error("Too many wins!")
    elseif sum(b) == 1 && !any(x == 0 for x in b)
        :draw
    elseif sum(b) ∉ (0, 1)
        error("Must take turns in order")
    else
        :ongoing
    end
end
```

glennj commented:
Do we want a test with an empty board? Would that state be ongoing?

Do we want to group the cases into scenarios?

SaschaMann commented:
> Do we want a test with an empty board? Would that state be ongoing?

I'm happy to add it as ongoing or leave it as undefined. I don't think it really makes a difference for any solutions.

> Do we want to group the cases into scenarios?

What grouping do you have in mind? Groups based on game state?

glennj commented:
Yes, that's what I was thinking.

SaschaMann commented:
@glennj I've grouped them based on game state now.

@petertseng Could you please check that I've configured the array formatter correctly?

SaschaMann commented:
@exercism/reviewers

petertseng reviewed (commented):
Yes, I confirm that this is the correct configuration for the array formatter.

The reason I haven't Approved, in the GitHub sense of the word, is just that I haven't reviewed the other contents, but you don't need to wait for me to do that if there are others who are doing that. I just wanted to quickly get a confirmation to you that that configuration was correct.

petertseng commented on exercises/state-of-tic-tac-toe/canonical-data.json:
```diff
@@ -0,0 +1,373 @@
+          "property": "gamestate",
+          "input": {
+            "board": [
+              "XXX",
+              "XOO",
+              "XOO"
+            ]
+          },
+          "expected": {
+            "error": "Impossible board: game should have ended after X won"
+          }
```
It is possible to reach this board from a legal game. X simply plays in the upper-left corner as the very last move. Therefore, it feels uncomfortable to call this board impossible. I understand that this is not rational play, but if this exercise is concerned with possible/impossible, then we have to consider all possible avenues through with a board can be reached, not only the rational ones.

petertseng commented on exercises/state-of-tic-tac-toe/canonical-data.json:
```diff
@@ -0,0 +1,373 @@
+          "description": "Invalid board",
+          "property": "gamestate",
+          "input": {
+            "board": [
+              "XOX",
+              "OXO",
+              "XOX"
+            ]
+          },
+          "expected": {
+            "error": "Impossible board: game should have ended after X won"
+          }
```
this board is also possible - X plays in the centre as the very last move.

ee7 commented on exercises/state-of-tic-tac-toe/description.md:
```diff
@@ -0,0 +1,69 @@
+If a board meets the following conditions, it is invalid:
+
+- The players took turns out of order (remember that `X` starts).
+- The game was played after it already ended, i.e. more than one row/column/diagonal contains the same symbol thrice.
```
(As an extension of petertseng's comments, we'd need a change here too):

"more than one row/column/diagonal contains the same symbol thrice" does not imply "the game was played after it already ended". So the former should not be a sufficient condition for "invalid".

From the "impossible board" cases, where `X` plays in the center as the final move:

```json
"board": [
  "XOX",
  "OXO",
  "XOX"
]
```

There are two diagonals that contain `X` thrice, but the game was not played after it already ended.

SaschaMann commented:
That's a really good point, I missed that you could win "twice" with one move. Fixed in de82efb, I moved them into wins because I think they could be interesting edge cases.

petertseng commented on exercises/state-of-tic-tac-toe/description.md:
````diff
@@ -0,0 +1,69 @@
+### Ongoing game
+
+```text
+   |   |   
+ X |   |   
````
```suggestion
   |   |
 X |   |
```

petertseng commented on exercises/state-of-tic-tac-toe/description.md:
```diff
@@ -0,0 +1,69 @@
+   |   |   
+ X |   |   
+___|___|___
+   |   |   
+   | X | O 
```
```suggestion
   |   |
   | X | O
```

petertseng commented on exercises/state-of-tic-tac-toe/description.md:
```diff
@@ -0,0 +1,69 @@
+   | X | O 
+___|___|___
+   |   |
+ O | X |  
```
```suggestion
 O | X |
```

petertseng commented on exercises/state-of-tic-tac-toe/description.md:
````diff
@@ -0,0 +1,69 @@
+### Draw
+
+```text
+   |   |   
+ X | O | X 
````
```suggestion
   |   |
 X | O | X
```

petertseng commented on exercises/state-of-tic-tac-toe/description.md:
```diff
@@ -0,0 +1,69 @@
+   |   |   
+ X | O | X 
+___|___|___
+   |   |   
+ X | X | O 
```
```suggestion
   |   |
 X | X | O
```

petertseng commented on exercises/state-of-tic-tac-toe/description.md:
````diff
@@ -0,0 +1,69 @@
+### Win
+
+```text
+   |   |   
+ X | X | X 
````
```suggestion
   |   |
 X | X | X
```

petertseng commented on exercises/state-of-tic-tac-toe/description.md:
```diff
@@ -0,0 +1,69 @@
+   |   |   
+ X | X | X 
+___|___|___
+   |   |   
+   | O | O 
```
```suggestion
   |   |
   | O | O
```

petertseng commented on exercises/state-of-tic-tac-toe/description.md:
```diff
@@ -0,0 +1,69 @@
+   | O | O 
+___|___|___
+   |   |
+   |   |  
```
```suggestion
   |   |
```

petertseng reviewed (approved):
while I'm making Suggestions (in the GitHub sense of the word) to remove trailing whitespace, they're not required. If we really want to avoid that we should probably have something that automatically checks for it.

SaschaMann commented:
> If we really want to avoid that we should probably have something that automatically checks for it.

I thought prettier did it but perhaps it's ignored in code blocks or something 🤷

kotp commented on exercises/state-of-tic-tac-toe/canonical-data.json:
```diff
@@ -0,0 +1,369 @@
+      "cases": [
+        {
+          "uuid": "b42ed767-194c-4364-b36e-efbfb3de8788",
+          "description": "Draw",
```
Should the descriptions be unique?  Is this being used for test functionality (function names, etc?) when generated?

SaschaMann replied:
That's not part of the spec and therefore would be an issue the generators would have to fix.

That said, the descriptions don't describe _why_ the test exists all that well, so if someone wants to improve those, a follow-up PR would be appreciated. The description is mutable and I'd rather get this merged and improved later on according to the usual optimistic merging strategy.

kotp commented on exercises/state-of-tic-tac-toe/description.md:
```diff
@@ -0,0 +1,69 @@
+
+If a board meets the following conditions, it is invalid:
+
+- The players took turns out of order (remember that `X` starts).
+- The game was played after it already ended.
+
```
How do we test for out of order?  Are there tests for out of order?  I only noticed snapshots and any snapshot that has equal or one less "O" than "X" can not indicate play order on its own.  I also do not see an example in "Examples" that shows what this might look like.

SaschaMann replied:
>  I only noticed snapshots and any snapshot that has equal or one less "O" than "X" can not indicate play order on its own.

A board that has more Os than Xs is impossible to reach when the right turn order is followed. We only care if the snapshot is a valid board, i.e. reachable through legal turns.

kotp replied:
So nothing to clarify about this in the description?  Line 24 mentions it, and I looked through and did not notice an example of an "out of order" state.  But does this matter?  Are we testing for this?  (Yes, a board that has more than 2 less "O"'s than "X" would an invalid state, etc.

SaschaMann replied:
There are two tests for it towards the end of the file. I'll add an example.

SaschaMann replied:
@kotp does [2f1e126](https://github.com/exercism/problem-specifications/pull/1991/commits/2f1e126d1ceb9bb1901fbfc94e4b29b6e5d0bb77) clear things up?

kotp replied:
Ah, yes, that works.

SaschaMann commented:
Thanks for the reviews :)
